File: ingest/go/pubSub.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
)

var (
	RabbitMqURL  = getEnv("RABBITMQ_URL", "amqp://guest:guest@localhost:5672/") // "amqp://guest:guest@192.168.1.205:5672/"
	MaxBatchSize = getEnvAsInt("MAX_BATCH_SIZE", 1500)                          // 100 kb
	BatchTimeout = getEnvAsDuration("BATCH_TIMEOUT", 10*time.Second)
	MaxRetries   = getEnvAsInt("MAX_RETRIES", 3)
	RetryDelay   = getEnvAsDuration("RETRY_DELAY", 500*time.Millisecond)
)

type PubSub struct {
	conn          *amqp.Connection
	ch            *amqp.Channel
	sessionID     string
	sessionTime   time.Time
	mu            sync.Mutex
	batchesLoaded int
	ctx           context.Context
	cancel        context.CancelFunc
	wg            sync.WaitGroup
	errorCh       chan error
	pointsBuffer  []byte
	bufferSize    int
	maxBufferSize int
}

func failOnError(err error, msg string) {
	if err != nil {
		log.Panicf("%s: %s", msg, err)
	}
}

func newPubSub(sessionId string, sessionTime time.Time) *PubSub {
	log.Printf("Connecting to RabbitMQ at %s", RabbitMqURL)

	conn, err := amqp.Dial(RabbitMqURL)
	failOnError(err, "Failed to connect to RabbitMQ")

	ch, err := conn.Channel()
	failOnError(err, "failed to open channel")

	err = ch.ExchangeDeclare("telemetry_topic", "topic", true, false, false, false, nil)
	failOnError(err, "Failed to declare exchange")

	q, err := ch.QueueDeclare(
		"telemetry_queue",
		true,
		false,
		false,
		false,
		nil,
	)
	failOnError(err, "Failed to declare queue")

	err = ch.QueueBind(
		q.Name,
		"telemetry.#",
		"telemetry_topic",
		false,
		nil,
	)
	failOnError(err, "Failed to bind queue")

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)

	return &PubSub{
		conn:          conn,
		ch:            ch,
		sessionID:     sessionId,
		sessionTime:   sessionTime,
		ctx:           ctx,
		cancel:        cancel,
		errorCh:       make(chan error, 100),
		pointsBuffer:  make([]byte, 0, MaxBatchSize),
		maxBufferSize: MaxBatchSize,
	}
}

func getFloatValue(record map[string]interface{}, key string) float64 {
	if val, ok := record[key]; ok {
		switch v := val.(type) {
		case float64:
			return v
		case string:
			f, err := strconv.ParseFloat(v, 64)
			if err == nil {
				return f
			}
			log.Printf("Float parse error for key %s: %v", key, err)
		case int:
			return float64(v)
		case int64:
			return float64(v)
		case float32:
			return float64(v)
		default:
			log.Printf("Unexpected type for key %s: %T", key, v)
		}
	}
	return 0.0
}

func getIntValue(record map[string]interface{}, key string) int {
	if val, ok := record[key]; ok {
		switch v := val.(type) {
		case int:
			return v
		case int64:
			return int(v)
		case float64:
			return int(v)
		case string:
			i, err := strconv.Atoi(v)
			if err == nil {
				return i
			}
			log.Printf("Integer parse error for key %s: %v", key, err)
		default:
			log.Printf("Unexpected type for key %s: %T", key, v)
		}
	}
	return 0
}

func (p *PubSub) Exec(data []map[string]interface{}) error {
	p.mu.Lock()
	defer p.mu.Unlock()

	if len(data) == 0 {
		return nil
	}

	for _, record := range data {
		lapID := getIntValue(record, "Lap")
		sessionTime := getFloatValue(record, "SessionTime")

		sessionNum := ""
		if val, ok := record["SessionNum"]; ok {
			sessionNum = fmt.Sprintf("%v", val)
		}

		trackName := ""
		if val, ok := record["trackDisplayShortName"]; ok {
			trackName = fmt.Sprintf("%v", val)
			trackName = strings.ReplaceAll(trackName, " ", "-")
		}

		trackID := ""
		if val, ok := record["trackID"]; ok {
			trackID = fmt.Sprintf("%v", val)
		}

		carID := ""
		if val, ok := record["PlayerCarIdx"]; ok {
			carID = fmt.Sprintf("%v", val)
		}

		tickTime := p.sessionTime.Add(time.Duration(sessionTime * float64(time.Second)))

		tick := map[string]interface{}{
			"lap_id":               fmt.Sprintf("%d", lapID),
			"speed":                getFloatValue(record, "Speed"),
			"lap_dist_pct":         getFloatValue(record, "LapDistPct"),
			"lap_current_lap_time": getFloatValue(record, "LapCurresntLapTime"),
			"session_id":           p.sessionID,
			"session_num":          sessionNum,
			"session_time":         sessionTime,
			"car_id":               carID,
			"track_name":           trackName,
			"track_id":             trackID,

			"steering_wheel_angle": getFloatValue(record, "SteeringWheelAngle"),
			"player_car_position":  getFloatValue(record, "PlayerCarPosition"),
			"velocity_x":           getFloatValue(record, "VelocityX"),
			"velocity_y":           getFloatValue(record, "VelocityY"),
			"velocity_z":           getFloatValue(record, "VelocityZ"),
			"fuel_level":           getFloatValue(record, "FuelLevel"),
			"throttle":             getFloatValue(record, "Throttle"),
			"brake":                getFloatValue(record, "Brake"),
			"rpm":                  getFloatValue(record, "RPM"),
			"lat":                  getFloatValue(record, "Lat"),
			"lon":                  getFloatValue(record, "Lon"),
			"gear":                 getIntValue(record, "Gear"),
			"alt":                  getFloatValue(record, "Alt"),
			"lat_accel":            getFloatValue(record, "LatAccel"),
			"long_accel":           getFloatValue(record, "LongAccel"),
			"vert_accel":           getFloatValue(record, "VertAccel"),
			"pitch":                getFloatValue(record, "Pitch"),
			"roll":                 getFloatValue(record, "Roll"),
			"yaw":                  getFloatValue(record, "Yaw"),
			"yaw_north":            getFloatValue(record, "YawNorth"),
			"tick_time":            tickTime.UTC(),
		}

		jsonData, err := json.Marshal(tick)
		if err != nil {
			log.Printf("Error marshaling point to JSON: %v", err)
		}

		if p.bufferSize == 0 {
			p.pointsBuffer = append(p.pointsBuffer, '[')
		} else {
			p.pointsBuffer = append(p.pointsBuffer, ',')
		}
		p.pointsBuffer = append(p.pointsBuffer, jsonData...)
		p.bufferSize++

		// Flush if buffer is full
		if p.bufferSize >= p.maxBufferSize {
			p.flushBuffer()
		}
	}

	p.batchesLoaded++
	log.Printf("Processed %d telemetry points in session %s (batch %d, buffer size: %d)",
		len(data), p.sessionID, p.batchesLoaded, p.bufferSize)

	return nil
}

func (p *PubSub) flushBuffer() {
	if p.bufferSize == 0 {
		return
	}

	p.pointsBuffer = append(p.pointsBuffer, ']')

	err := p.ch.PublishWithContext(p.ctx, "telemetry_topic", "telemetry.ticks", true, false, amqp.Publishing{
		ContentType: "text/plain",
		Body:        p.pointsBuffer,
	})

	failOnError(err, "Failed to publish message")

	p.pointsBuffer = p.pointsBuffer[:0]
	p.bufferSize = 0
}

func (p *PubSub) Close() error {
	p.cancel()
	p.wg.Wait()

	if p.ch != nil {
		p.ch.Close()
	}

	if p.conn != nil {
		p.conn.Close()
	}

	log.Println("Closed RabbitMQ connection.")
	return nil
}

func (p *PubSub) Loaded() int {
	return p.batchesLoaded
}

func getEnv(key, fallback string) string {
	if value, exists := os.LookupEnv(key); exists {
		return value
	}
	return fallback
}

func getEnvAsInt(key string, fallback int) int {
	if value, exists := os.LookupEnv(key); exists {
		if intVal, err := strconv.Atoi(value); err == nil {
			return intVal
		}
	}
	return fallback
}

func getEnvAsDuration(key string, fallback time.Duration) time.Duration {
	if value, exists := os.LookupEnv(key); exists {
		if duration, err := time.ParseDuration(value); err == nil {
			return duration
		}
	}
	return fallback
}

File: ingest/dotnet/src/PubSub/PubSub.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using Microsoft.Extensions.Logging;
using SVappsLAB.iRacingTelemetrySDK;
using RabbitMQ.Client;

namespace ingest.PubSub;

class BufferedPubSub : IDisposable
{
    private IConnection? _connection;
    private IChannel? _channel;
    private readonly object _lockObject = new object();
    private bool _isConnected = false;
    private bool _disposed = false;

    private readonly int _maxBatchSize;
    private readonly int _maxBatchBytes;
    private readonly TimeSpan _flushInterval;

    private readonly List<object> _buffer = new();
    private readonly Timer _flushTimer;
    private int _currentBufferBytes;

    private long _totalPointsBuffered;
    private long _totalBatchesSent;
    private DateTime _lastFlush = DateTime.UtcNow;

    private readonly CancellationTokenSource _cancellationTokenSource = new();

    public BufferedPubSub(
        int maxBatchSize = 1000,
        int maxBatchBytes = 250000,
        TimeSpan? flushInterval = null)
    {
        _maxBatchSize = maxBatchSize;
        _maxBatchBytes = maxBatchBytes;
        _flushInterval = flushInterval ?? TimeSpan.FromMilliseconds(50);

        _flushTimer = new Timer(FlushTimerCallback, null, _flushInterval, _flushInterval);
        
        Console.WriteLine($"BufferedPubSub initialized: maxBatch={_maxBatchSize}, maxBytes={_maxBatchBytes}, flushInterval={_flushInterval.TotalMilliseconds}ms");
    }

    public async Task InitializeAsync()
    {
        if (_isConnected || _disposed) return;

        lock (_lockObject)
        {
            if (_isConnected || _disposed) return;

            try
            {
                var factory = new ConnectionFactory();
                factory.Uri = new Uri("amqp://guest:guest@localhost:5672/");

                _connection = factory.CreateConnectionAsync().Result;
                _channel = _connection.CreateChannelAsync(null).Result;

                _isConnected = true;
                Console.WriteLine("RabbitMQ connection established and reused for all messages");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Failed to initialize RabbitMQ connection: {ex.Message}");
                _isConnected = false;
                throw;
            }
        }
    }

    public async void Publish(ILogger logger, TelemetryData data, string trackName = "", string trackId = "", int sessionId = 0)
    {
        if (_disposed || _cancellationTokenSource.Token.IsCancellationRequested) 
        {
            logger.LogWarning("Publish called after disposal or cancellation");
            return;
        }

        try
        {
            await InitializeAsync();

            if (_channel == null)
            {
                Console.WriteLine("RabbitMQ channel not available");
                return;
            }

            var tick = new
            {
                lap_id = data.Lap.ToString(),
                speed = data.Speed,
                lap_dist_pct = data.LapDistPct,
                session_id = sessionId,
                session_num = data.SessionNum,
                session_time = data.SessionTime,
                car_id = data.PlayerCarIdx,
                track_name = trackName,
                track_id = trackId,
                worker_id = 1,
                steering_wheel_angle = data.SteeringWheelAngle,
                player_car_position = data.PlayerCarPosition,
                velocity_x = data.VelocityX,
                velocity_y = data.VelocityY,
                velocity_z = data.VelocityZ,
                fuel_level = data.FuelLevel,
                throttle = data.Throttle,
                brake = data.Brake,
                rpm = data.RPM,
                lat = data.Lat,
                lon = data.Lon,
                gear = data.Gear,
                alt = data.Alt,
                lat_accel = data.LatAccel,
                long_accel = data.LongAccel,
                vert_accel = data.VertAccel,
                pitch = data.Pitch,
                roll = data.Roll,
                yaw = data.Yaw,
                yaw_north = data.YawNorth,
                voltage = data.Voltage,
                lapLastLapTime = data.LapLastLapTime,
                waterTemp = data.WaterTemp,
                lapDeltaToBestLap = data.LapDeltaToBestLap,
                lapCurrentLapTime = data.LapCurrentLapTime,
                lFpressure = data.LFpressure,
                rFpressure = data.RFpressure,
                lRpressure = data.LRpressure,
                rRpressure = data.RRpressure,
                lFtempM = data.LFtempM,
                rFtempM = data.RFtempM,
                lRtempM = data.LRtempM,
                rRtempM = data.RRtempM,
                tick_time = DateTime.UtcNow
            };

            await BufferTick(tick, sessionId);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error in buffered publish: {ex}");
            _isConnected = false;
        }
    }

    private async Task BufferTick(object tick, int sessionId)
    {
        if (_disposed || _cancellationTokenSource.Token.IsCancellationRequested) return;

        bool shouldFlush = false;
        
        lock (_lockObject)
        {
            if (_disposed) return;

            _buffer.Add(tick);
            _currentBufferBytes += EstimateTickSize(tick);
            _totalPointsBuffered++;

            shouldFlush = _buffer.Count >= _maxBatchSize || _currentBufferBytes >= _maxBatchBytes;
            
            if (_buffer.Count % 500 == 0)
            {
                Console.WriteLine($"Buffer status: {_buffer.Count}/{_maxBatchSize} points, {_currentBufferBytes}/{_maxBatchBytes} bytes");
            }
        }

        if (shouldFlush)
        {
            Console.WriteLine($"Triggering immediate flush: {_buffer.Count} points, {_currentBufferBytes} bytes");
            await FlushBuffer(sessionId);
        }
    }

    private void FlushTimerCallback(object? state)
    {
        if (_disposed || _cancellationTokenSource.Token.IsCancellationRequested) return;

        _ = Task.Run(async () =>
        {
            try
            {
                await FlushBuffer(1);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in timer-triggered flush: {ex.Message}");
            }
        }, _cancellationTokenSource.Token);
    }

    private async Task FlushBuffer(int sessionId)
    {
        if (_disposed || _cancellationTokenSource.Token.IsCancellationRequested) return;

        List<object> dataToFlush;

        lock (_lockObject)
        {
            if (_disposed || _buffer.Count == 0)
                return;

            dataToFlush = new List<object>(_buffer);
            _buffer.Clear();
            _currentBufferBytes = 0;
        }

        try
        {
            await SendBatchToRabbitMQ(dataToFlush, sessionId);

            lock (_lockObject)
            {
                _totalBatchesSent++;
                _lastFlush = DateTime.UtcNow;
            }

            if (_totalBatchesSent % 100 == 0)
            {
                var elapsed = DateTime.UtcNow - _lastFlush;
                if (elapsed.TotalSeconds > 0)
                {
                    var rate = _totalPointsBuffered / elapsed.TotalSeconds;
                    Console.WriteLine($"Flush metrics: {_totalPointsBuffered} points in {_totalBatchesSent} batches ({rate:F0} points/sec)");
                }
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"ERROR flushing buffer to RabbitMQ: {ex.Message}");
        }
    }

    private async Task SendBatchToRabbitMQ(List<object> batch, int sessionId)
    {
        if (batch.Count == 0 || _channel == null || _disposed) return;

        try
        {
            Dictionary<string, object?> headers = new Dictionary<string, object?>();
            headers.Add("worker_id", 1);
            headers.Add("batch_size", batch.Count);
            headers.Add("session_id", sessionId);

            string jsonString = JsonSerializer.Serialize(batch, new JsonSerializerOptions
            {
                WriteIndented = false
            });

            byte[] messageBodyBytes = System.Text.Encoding.UTF8.GetBytes(jsonString);

            var props = new BasicProperties();
            props.ContentType = "application/json";
            props.DeliveryMode = DeliveryModes.Transient;
            props.Timestamp = new AmqpTimestamp(DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
            props.Headers = headers;

            await _channel.BasicPublishAsync("telemetry_topic", "telemetry.ticks", false, props, messageBodyBytes,
                _cancellationTokenSource.Token);

            Console.WriteLine($"Sent batch of {batch.Count} points ({messageBodyBytes.Length} bytes) to RabbitMQ");
        }
        catch (OperationCanceledException)
        {
            Console.WriteLine("Batch sending was cancelled");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error sending batch to RabbitMQ: {ex}");
            throw;
        }
    }

    private static int EstimateTickSize(object tick)
    {
        return 600;
    }

    public async Task ForceFlush(int sessionId = 1)
    {
        if (_disposed) return;
        
        Console.WriteLine("Force flushing remaining data...");
        await FlushBuffer(sessionId);
        Console.WriteLine("Force flush completed");
    }

    public void Dispose()
    {
        if (_disposed) return;
        
        _disposed = true;
        
        Console.WriteLine("Starting BufferedPubSub disposal...");
        
        _cancellationTokenSource.Cancel();
        
        try
        {
            _flushTimer?.Dispose();
            Console.WriteLine("Timer disposed");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error disposing timer: {ex.Message}");
        }
        
        try
        {
            var flushTask = ForceFlush();
            if (flushTask.Wait(TimeSpan.FromSeconds(3)))
            {
                Console.WriteLine("Final flush completed successfully");
            }
            else
            {
                Console.WriteLine("Final flush timed out");
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error during final flush: {ex.Message}");
        }
        
        lock (_lockObject)
        {
            try
            {
                _channel?.CloseAsync().Wait(TimeSpan.FromSeconds(2));
                _channel?.Dispose();
                Console.WriteLine("Channel closed and disposed");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error disposing channel: {ex.Message}");
            }
            
            try
            {
                _connection?.CloseAsync().Wait(TimeSpan.FromSeconds(2));
                _connection?.Dispose();
                Console.WriteLine("Connection closed and disposed");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error disposing connection: {ex.Message}");
            }
            
            _isConnected = false;
        }
        
        try
        {
            _cancellationTokenSource.Dispose();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error disposing cancellation token source: {ex.Message}");
        }
        
        Console.WriteLine($"BufferedPubSub disposed. Final stats: {_totalPointsBuffered} points, {_totalBatchesSent} batches");
    }
}